sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamTableFunctionScanRel.java

```diff
@@ -0,0 +1,157 @@
+  public PTransform<PCollectionList<Row>, PCollection<Row>> buildPTransform() {
+    return new Transform();
+  }
+
+  private class Transform extends PTransform<PCollectionList<Row>, PCollection<Row>> {
+
+    @Override
+    public PCollection<Row> expand(PCollectionList<Row> input) {
+      checkArgument(
+          input.size() == 1,
+          "Wrong number of inputs for %s: %s",
+          BeamTableFunctionScanRel.class.getSimpleName(),
+          input);
+      checkArgument(
+          ((RexCall) getCall()).getOperator().getName().equals("TUMBLE"),
+          "Only support TUMBLE table-valued function");
+      RexCall call = ((RexCall) getCall());
+      RexInputRef wmCol = (RexInputRef) call.getOperands().get(1);
+      PCollection<Row> upstream = input.get(0);
+      Schema outputSchema = CalciteUtils.toSchema(getRowType());
+      return upstream
+          .apply(
+              ParDo.of(
+                  new FixedWindowDoFn(
+                      FixedWindows.of(durationParameter(call.getOperands().get(2))),
+                      wmCol.getIndex(),
+                      outputSchema)))
+          .setRowSchema(outputSchema);
+    }
+  }
+
+  private Duration durationParameter(RexNode node) {
+    return Duration.millis(longValue(node));
+  }
+
+  private long longValue(RexNode operand) {
+    if (operand instanceof RexLiteral) {
+      return ((Number) RexLiteral.value(operand)).longValue();
+    } else {
+      throw new IllegalArgumentException(String.format("[%s] is not valid.", operand));
+    }
+  }
+
+  private static class FixedWindowDoFn extends DoFn<Row, Row> {
+    private int windowFieldIndex;
+    private FixedWindows windowFn;
+    private Schema outputSchema;
+
+    public FixedWindowDoFn(FixedWindows windowFn, int windowFieldIndex, Schema schema) {
+      this.windowFn = windowFn;
+      this.windowFieldIndex = windowFieldIndex;
+      this.outputSchema = schema;
+    }
+
+    @ProcessElement
+    public void processElement(ProcessContext c) {
+      Row row = c.element();
+      IntervalWindow window = windowFn.assignWindow(row.getDateTime(windowFieldIndex).toInstant());
+      Row.Builder builder = Row.withSchema(outputSchema);
+      builder.addValues(row.getValues());
+      // TODO: add window_start and window_end;
+      builder.addValue(window.start());
+      builder.addValue(window.end());
+      c.output(builder.build());
+    }
+  }
+
+  @Override
+  public NodeStats estimateNodeStats(RelMetadataQuery mq) {
+    return BeamSqlRelUtils.getNodeStats(getInput(0), mq);
+  }
+
+  @Override
+  public BeamCostModel beamComputeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
+    NodeStats inputEstimates = BeamSqlRelUtils.getNodeStats(getInput(0), mq);
+
+    final double rowSize = getRowType().getFieldCount();
+    final double cpu = inputEstimates.getRowCount() * inputEstimates.getRowCount() * rowSize;
+    final double cpuRate = inputEstimates.getRate() * inputEstimates.getWindow() * rowSize;
```

Comment:
Can you explain the logic here? It may be worth adding a comment to describe the calculation.

For example, how does cpu = row_count^2 * row_size ? It isn't obvious to me.

Reply:
I believe it is carried over from a class in BeamSQL. Such computation was a part of a intern project before about SQL cost based optimization.

Let me do some research and see how it should work.

Reply:
Yeah the cpu computation is wrong (I copied it from SortRel).

The reasonable computation is `cpu = inputEstimates.getRowCount() * rowSize * function_per_row_cost_factor;`. Because we don't have a reasonable stats for `function_per_row_cost_factor`, the formula is simplified as row count times row size.